packages/editor/src/hooks/custom-sources-backwards-compatibility.js

```diff
@@ -1,76 +1,95 @@
 import { addFilter } from '@wordpress/hooks';
 
-const EMPTY_OBJECT = {};
-function useMetaAttributeSource( name, _attributes, _setAttributes ) {
-	const { attributes: attributeTypes = EMPTY_OBJECT } =
-		getBlockType( name ) || EMPTY_OBJECT;
-	let [ attributes, setAttributes ] = [ _attributes, _setAttributes ];
+/** @typedef {import('@wordpress/compose').WPHigherOrderComponent} WPHigherOrderComponent */
+
+/** @typedef {import('@wordpress/blocks').WPBlockSettings} WPBlockSettings */
+
+/**
+ * Object mapping an attribute key to its corresponding meta property.
+ *
+ * @typedef {Object<string,string>} WPMetaAttributeMapping
+ */
 
-	if ( Object.values( attributeTypes ).some( ( type ) => type.source === 'meta' ) ) {
-		// eslint-disable-next-line react-hooks/rules-of-hooks
-		const type = useSelect( ( select ) => select( 'core/editor' ).getCurrentPostType(), [] );
-		// eslint-disable-next-line react-hooks/rules-of-hooks
-		const [ meta, setMeta ] = useEntityProp( 'postType', type, 'meta' );
+/**
+ * Given a meta attribute mapping, returns a new higher-order component which
+ * manages attributes merging and updates to consume from and mirror to the
+ * associated entity record.
+ *
+ * @param {WPMetaAttributeMapping} metaKeys Meta attribute mapping.
+ *
+ * @return {WPHigherOrderComponent} Higher-order component.
+ */
+const createWithMetaAttributeSource = ( metaKeys ) => createHigherOrderComponent(
+	( BlockEdit ) => ( { attributes, setAttributes, ...props } ) => {
+		const postType = useSelect( ( select ) => select( 'core/editor' ).getCurrentPostType() );
+		const [ meta, setMeta ] = useEntityProp( 'postType', postType, 'meta' );
 
-		// eslint-disable-next-line react-hooks/rules-of-hooks
-		attributes = useMemo(
+		const mergedAttributes = useMemo(
 			() => ( {
-				..._attributes,
-				...Object.keys( attributeTypes ).reduce( ( acc, key ) => {
-					if ( attributeTypes[ key ].source === 'meta' ) {
-						acc[ key ] = meta[ attributeTypes[ key ].meta ];
-					}
-					return acc;
-				}, {} ),
+				...attributes,
+				...mapValues( metaKeys, ( metaKey ) => meta[ metaKey ] ),
 			} ),
-			[ attributeTypes, meta, _attributes ]
+			[ attributes, meta ]
 		);
 
-		// eslint-disable-next-line react-hooks/rules-of-hooks
-		setAttributes = useCallback(
-			( ...args ) => {
-				Object.keys( args[ 0 ] ).forEach( ( key ) => {
-					if ( attributeTypes[ key ].source === 'meta' ) {
-						setMeta( { [ attributeTypes[ key ].meta ]: args[ 0 ][ key ] } );
+		return (
+			<BlockEdit
+				attributes={ mergedAttributes }
+				setAttributes={ ( nextAttributes ) => {
+					const nextMeta = mapKeys(
+						// Filter to intersection of keys between the updated
+						// attributes and those with an associated meta key.
+						pickBy( nextAttributes, ( value, key ) => metaKeys[ key ] ),
+
+						// Rename the keys to the expected meta key name.
+						( value, attributeKey ) => metaKeys[ attributeKey ],
```

Comment:
> A map to `metaKeys[ key ]` and filtering off `! metaKeys[ key ]` might be clearer.

Hm, I'm not really sure what you're suggesting here. I've tried a few different approaches based on what I think you're proposing, but a few issues to account for:

- Mapping keys first means we have to deal with the fact that not all of the next attributes will have an associated meta key (non-meta-sourced attributes)
- Dealing with the filter after means we have to be sure that the `nextAttributes` includes the keys, not necessarily that it's truthy (e.g. need to allow the meta to be set to an empty string)

Reply:
I think I was thinking of something like this:

```js
mapValues(
	filter(
		mapValues( nextAttributes, ( value, key ) => metaKeys[ key ] ),
		( value ) => value
	),
	( value, key ) => nextAttributes[ key ]
);
```

But, it's not actually simpler :rofl:

Reply:
Not sure that would work, either, since `filter` returns an array (i.e. numeric keys). I guess it'd need to be a `pickBy`, but like you mention, not really a simplification to go that route.

Reply:
Right, I thought you could also filter objects into objects, but that is a `pickBy`.